mogp_emulator/GaussianProcessGPU.py

```diff
@@ -0,0 +1,421 @@
+
+        Returns a string indicating how the nugget parameter is treated, either ``"adaptive"``,
+        ``"fit"``, or ``"fixed"``. This is automatically set when changing the ``nugget``
+        property.
+
+        :returns: Current nugget fitting method
+        :rtype: str
+        """
+        return self._nugget_type.__str__().split(".")[1]
+
+    @property
+    def nugget(self):
+        """
+        See :func:`mogp_emulator.GaussianProcess.GaussianProcess.nugget`
+        """
+        return self._nugget
+
+    @nugget.setter
+    def nugget(self, nugget):
+        if not isinstance(nugget, (str, float)):
+            try:
+                nugget = float(nugget)
+            except TypeError:
+                raise TypeError("nugget parameter must be a string or a non-negative float")
+
+        if isinstance(nugget, str):
+            if nugget == "adaptive":
+                self._nugget_type = LibGPGPU.nugget_type(0)
+            elif nugget == "fit":
+                self._nugget_type = LibGPGPU.nugget_type(1)
+            else:
+                raise ValueError("nugget must be a float set to 'adaptive', 'fit', or 'fixed'")
+            self._nugget = None
+        else:
+            if nugget < 0.:
+                raise ValueError("nugget parameter must be non-negative")
+            self._nugget_type = LibGPGPU.nugget_type(2) #fixed
+            self._nugget = float(nugget)
+
+    @property
+    def theta(self):
+        """
+        Returns emulator hyperparameters
+        see
+        :func:`mogp_emulator.GaussianProcess.GaussianProcess.theta`
+
+        :type theta: ndarray
+        """
+        if not self._densegp_gpu.theta_fit_status():
+            return None
+        theta = np.zeros(self.n_params)
+        self._densegp_gpu.get_theta(theta)
+        return theta
+
+    @theta.setter
+    def theta(self, theta):
+        """
+        Fits the emulator and sets the parameters (property-based setter
+        alias for ``fit``)
+
+        See :func:`mogp_emulator.GaussianProcess.GaussianProcess.theta`
+
+        :type theta: ndarray
+        :returns: None
+        """
+        if theta is None:
+            self._densegp_gpu.reset_theta_fit_status()
+            self.current_logpost = None
+        else:
+            self.fit(theta)
+
+    @property
+    def L(self):
+        """
+        Return the lower triangular Cholesky factor.
+
+        :returns: np.array
+        """
+        result = np.zeros((self.n, self.n))
+        self._densegp_gpu.get_cholesky_lower(result)
+        return np.tril(result.transpose())
+
+    def get_K_matrix(self):
+        """
+        Returns current value of the inverse covariance matrix as a numpy array.
+
+        Does not include the nugget parameter, as this is dependent on how the 
+        nugget is fit.
+        """
+        result = np.zeros((self.n, self.n))
+        self._densegp_gpu.get_invQ(result)
+        return np.linalg.inv(result)
```

Comment:
Yes, I think you're right - in fact the approach here will give the wrong K-matrix for a non-zero nugget, as invQ includes the nugget.
In the C++ we now cache the nuggetless K matrix in addition to invQ, and return that here.  This may not be the best long-term approach if we want to minimize memory usage on the GPU - in that case we could instead recompute it on the fly.   We can perhaps revisit that later.